compiler/il/OMRSymbol.hpp

```diff
@@ -494,17 +510,34 @@ class OMR_EXTENSIBLE Symbol
        */
       KindMask                  = 0x00000700,
 
-
       IsInGlobalRegister        = 0x00000800,
       Const                     = 0x00001000,
-      Volatile                  = 0x00002000,
-      InitializedReference      = 0x00004000,
+
+      /**
+       * Memory access ordering semantics flags
+       */
```

Comment:
it looks like you are mixing up two types of ordering: 1) the order instructions are laid down in memory (more closely related to instruction-dispatching order. that might be what you meant by program order. relevant to compiler optimization.);  2)  the order in which memory accesses are executed/observed (more closely related to instruction-issuing or cache RC-machine actioning. need of memory-barriers to enforce a certain order.).

by a quick glimpse of the code, i have a high-level comment for you to consider further:  you seemed changing a lot of places querying 2) above into using querying 1) above.  is that optimal or even possibly having correctness-implication?

Reply:
`Opaque` semantics means that the execution order of an access to the symbol will match the program order as observed by the executing thread.

Reply:
on weakly-consistent machine (e.g. POWER), you don't have guarantee of the execution order at all without memory barriers.  i.e. program order is meaning-less unless we are talking about accesses to the same location.

Reply:
From the Java `VarHandle` description `Opaque` semantics description:

> Opaque operations are bitwise atomic and coherently ordered with respect to accesses to the same variable.

If I'm understanding those semantics correctly, for `Opaque` we only need to ensure that accesses to the same variable/address are executed in the same order (as seen by the executing thread) with respect to each other as they are laid down in the program order. If I'm understanding the weakly-consistent machine memory model correctly, the address dependency between the accesses should be enough to ensure this order without memory barriers.